07_boolean_2/main.go

```diff
@@ -15,56 +15,262 @@ E.g.:
+//   - If the integer is odd, return "longer and odd"
+//
+// - If the string is shorter than the integer, return "shorter"
+func strIntCheck(str string, num int) {
+	// var stringLength = len(str)
+	// var integerLength = len(num)
+	if str == "" {
+		fmt.Println("empty")
+	} else if len(str) > num {
+		if len(str)%2 == 0 {
+			fmt.Println("longer and even")
+		} else {
+			fmt.Println("longer and odd")
+		}
+	} else {
+		fmt.Println("shorter")
+	}
+}
+
+// TODO: Write a function that takes 5 integers (grades) and returns the
+// following:
+// - If any of the grades is less than 0 or greater than 100, return
+// "invalid grade".
+// - If the average of the grades is greater than or equal to 90
+//   - If no grade is below 80, return "A+"
+//   - If no grade is below 70, return "A"
+//   - If any grade is below 70, return "A-"
+//
+// - If the average of the grades is greater than or equal to 80, return
+// "B"
+//   - If no grade is below 80, return "B+"
+//   - If no grade is below 70, return "B"
+//   - If any grade is below 60, return "B-"
+//
+// - If the average of the grades is greater than or equal to 70, return
+// "C"
+//
+//   - If no grade is below 70, return "C+"
+//
+//   - If no grade is below 60, return "C"
+//
+//   - If any grade is below 60, return "C-"
+
+// if grade.. > 100 || grade.. < 0 return invalid
+func gradeValid(grade1 int, grade2 int, grade3 int, grade4 int, grade5 int) bool {
+	if (grade1 < 0 || grade1 > 100) || (grade2 < 0 || grade2 > 100) || (grade3 < 0 || grade3 > 100) || (grade4 < 0 || grade4 > 100) || (grade5 < 0 || grade5 > 100) {
+		return false
+	} else {
+		return true
+	}
+}
+
+// check if any grade is < x
+func anyGradeUnder(grade1 int, grade2 int, grade3 int, grade4 int, grade5 int, limit int) bool {
+	if grade1 < limit || grade2 < limit || grade3 < limit || grade4 < limit || grade5 < limit {
+		return true
+	} else {
+		return false
+	}
+}
+
+// check if all grades are >= x
+func allGradesOver(grade1 int, grade2 int, grade3 int, grade4 int, grade5 int, limit int) bool {
+	if grade1 >= limit && grade2 >= limit && grade3 >= limit && grade4 >= limit && grade5 >= limit {
+		return true
+	} else {
+		return false
+	}
+}
+
+// check if grade avg >= limit
+func gradeAvgOver(grade1 int, grade2 int, grade3 int, grade4 int, grade5 int, limit int) bool {
+	if (grade1+grade2+grade3+grade4+grade5)/5 >= limit {
+		return true
+	} else {
+		return false
+	}
+}
+
+func calcGrade(grade1 int, grade2 int, grade3 int, grade4 int, grade5 int) string {
+	if gradeValid(grade1, grade2, grade3, grade4, grade5) == false {
+		return "invalid grade"
+	}
+	if gradeAvgOver(grade1, grade2, grade3, grade4, grade5, 90) == true {
+		if allGradesOver(grade1, grade2, grade3, grade4, grade5, 80) == true {
+			return "A+"
+		} else if allGradesOver(grade1, grade2, grade3, grade4, grade5, 70) == true {
+			return "A"
+		} else if anyGradeUnder(grade1, grade2, grade3, grade4, grade5, 70) == true {
+			return "A-"
+		}
```

Comment:
Something that you're doing here that can be improved.

Each time you `return`, the execution of the program is stopped and goes back up to the parent function.

So if you `return` from an `if` statement, you don't need `else` or `else if`.

E.g. (don't worry about the logic itself, it's meaningless):

```go
func check1(x int, y int) bool {
	if x == y {
		return false
	} else {
		if x*2 > y {
			return false
		} else if x*3 > y {
			return true
		}
	}
	return false
}

func check2(x int, y int) bool {
	if x == y {
		return false
	}

	if x*2 > y {
		return false
	}
	
	if x*3 > y {
		return true
	}
	
	return false
}
```

See how `check1` has 2 levels of `if`/`else`, whereas the `check2` has only one. The second one reads _much_ better and is strictly equivalent.

Have a look at https://www.youtube.com/watch?v=CFRhGnuXG-4 for inspiration.